File: frontend/src/flowMermaid.js
const labelReplacements = new Map([
  ['"', "'"], ["\n", " "], ["\r", " "], ["[", "［"], ["]", "］"],
  ["{", "｛"], ["}", "｝"], ["(", "（"], [")", "）"], ["|", "｜"],
  ["<", "＜"], [">", "＞"], ["&", "＆"], ["#", "＃"],
]);

function safeLabel(value = "") {
  return String(value)
    .replace(/["\n\r[\]{}()|<>&#]/g, (character) => labelReplacements.get(character) || character)
    .replace(/\s+/g, " ")
    .trim();
}

function nodeAliases(flow) {
  const aliases = new Map();
  const used = new Set();
  (flow.nodes || []).forEach((node, index) => {
    let base = String(node.id || "").replace(/[^A-Za-z0-9_]/g, "_");
    if (!base || /^\d/.test(base)) base = `n_${base || index + 1}`;
    let alias = base;
    let suffix = 2;
    while (used.has(alias)) {
      alias = `${base}_${suffix}`;
      suffix += 1;
    }
    aliases.set(node.id, alias);
    used.add(alias);
  });
  return aliases;
}

export function productFlowToMermaid(flow) {
  const aliases = nodeAliases(flow);
  const lines = ["flowchart TD"];
  (flow.nodes || []).forEach((node) => {
    const id = aliases.get(node.id);
    const label = safeLabel(node.label);
    if (node.node_type === "start" || node.node_type === "end") lines.push(`${id}([${label}])`);
    else if (node.node_type === "decision") lines.push(`${id}{${label}}`);
    else lines.push(`${id}[${label}]`);
  });
  (flow.edges || []).forEach((edge) => {
    const source = aliases.get(edge.source);
    const target = aliases.get(edge.target);
    const label = safeLabel(edge.label);
    lines.push(label ? `${source} -->|${label}| ${target}` : `${source} --> ${target}`);
  });
  return lines.join("\n");
}

File: utils/flow_mermaid.py
"""Deterministic ProductFlow-to-Mermaid conversion."""

import re

from models.flow import ProductFlow


_LABEL_TRANSLATION = str.maketrans(
    {
        '"': "'",
        "\n": " ",
        "\r": " ",
        "[": "［",
        "]": "］",
        "{": "｛",
        "}": "｝",
        "(": "（",
        ")": "）",
        "|": "｜",
        "<": "＜",
        ">": "＞",
        "&": "＆",
        "#": "＃",
    }
)


def _safe_label(value: str) -> str:
    """Flatten and neutralize Mermaid delimiter characters in labels."""
    return " ".join(value.translate(_LABEL_TRANSLATION).split())


def _node_ids(flow: ProductFlow) -> dict[str, str]:
    """Create stable Mermaid-safe aliases while preserving valid IDs."""
    aliases: dict[str, str] = {}
    used: set[str] = set()
    for index, node in enumerate(flow.nodes, start=1):
        base = re.sub(r"[^A-Za-z0-9_]", "_", node.id)
        if not base or base[0].isdigit():
            base = f"n_{base or index}"
        alias = base
        suffix = 2
        while alias in used:
            alias = f"{base}_{suffix}"
            suffix += 1
        aliases[node.id] = alias
        used.add(alias)
    return aliases


def product_flow_to_mermaid(flow: ProductFlow) -> str:
    """Convert ProductFlow into deterministic top-to-bottom Mermaid text."""
    aliases = _node_ids(flow)
    lines = ["flowchart TD"]
    for node in flow.nodes:
        node_id = aliases[node.id]
        label = _safe_label(node.label)
        if node.node_type in {"start", "end"}:
            lines.append(f"{node_id}([{label}])")
        elif node.node_type == "decision":
            lines.append(f"{node_id}{{{label}}}")
        else:
            lines.append(f"{node_id}[{label}]")
    for edge in flow.edges:
        source = aliases[edge.source]
        target = aliases[edge.target]
        label = _safe_label(edge.label)
        lines.append(
            f"{source} -->|{label}| {target}"
            if label
            else f"{source} --> {target}"
        )
    return "\n".join(lines)
